cordapp-test-utils/test-utils/src/main/kotlin/net/corda/testutils/services/DbPersistenceService.kt

```diff
@@ -0,0 +1,139 @@
+                        JPA_JDBC_USER to "admin",
+                        JPA_JDBC_PASSWORD to "",
+                        DIALECT to HSQLDialect::class.java.name,
+                        HBM2DDL_AUTO to "create",
+                    )
+                )
+        }
+    }
+
+    override fun <T : Any> find(entityClass: Class<T>, primaryKey: Any): T? {
+        return emf.guard {
+            it.find(entityClass, primaryKey)
+        }
+    }
+
+    override fun <T : Any> find(entityClass: Class<T>, primaryKeys: List<Any>): List<T> {
+        emf.guard {
+            return primaryKeys.map { pk -> it.find(entityClass, pk) }
+        }
+    }
+
+    override fun <T : Any> findAll(entityClass: Class<T>): PagedQuery<T> {
+        emf.guard {
+            val query = it.createQuery("SELECT e FROM ${entityClass.simpleName} e")
+            val result = cast<List<T>>(query.resultList)
+                ?: throw java.lang.IllegalArgumentException("The result of the query was not an $entityClass")
+
+            return object : PagedQuery<T> {
+                override fun execute(): List<T>  = result
+
+                override fun setLimit(limit: Int): PagedQuery<T> { TODO("Not yet implemented") }
+
+                override fun setOffset(offset: Int): PagedQuery<T> { TODO("Not yet implemented") }
+            }
+        }
+    }
+
+    override fun <T : Any> merge(entity: T): T? {
+        emf.transaction {
+            return it.merge(entity)
+        }
+    }
+
+    override fun <T : Any> merge(entities: List<T>): List<T> {
+        emf.transaction { em ->
+            return entities.map { em.merge(it) }
+        }
+    }
+
+    override fun persist(entity: Any) {
+        emf.transaction {
+            it.persist(entity)
+        }
+    }
+
+    override fun persist(entities: List<Any>) {
+        emf.transaction { em ->
+            entities.forEach { em.persist(it) }
+        }
+    }
+
+    override fun <T : Any> query(queryName: String, entityClass: Class<T>): ParameterisedQuery<T> {
+        TODO("Not yet implemented")
+    }
+
+    override fun remove(entity: Any) {
+        emf.transaction {
+            it.remove(if (it.contains(entity)) { entity } else { it.merge(entity) })
+        }
+    }
+
+    override fun remove(entities: List<Any>) {
+        emf.transaction {
+            entities.forEach { entity ->
+                it.remove(if (it.contains(entity)) { entity } else { it.merge(entity) })
+            }
+        }
+    }
+}
+
+inline fun <R> EntityManagerFactory.transaction(block: (EntityManager) -> R): R {
+
+    val em = this.createEntityManager()
+    val t = em.transaction
+    t.begin()
+    return try {
+        block(em)
+    } catch (e: Exception) {
+        t.setRollbackOnly()
+        throw CordaPersistenceException(e.message ?: "Error in persistence", e)
+    } finally {
+        if (!t.rollbackOnly) {
+            t.commit()
+        } else {
+            t.rollback()
+        }
+    }
```

Comment:
You need to do something like: 
```suggestion
this.createEntityManager().use { em ->
    val t = em.transaction
    t.begin()
    return try {
        block(em)
    } catch (e: Exception) {
        t.setRollbackOnly()
        throw CordaPersistenceException(e.message ?: "Error in persistence", e)
    } finally {
        if (!t.rollbackOnly) {
            t.commit()
        } else {
            t.rollback()
        }
    }
}
```

To ensure the EM is closed when you're finished.
As it stands, this will leak.